homeassistant/components/fritz/common.py

```diff
@@ -224,29 +229,95 @@ def scan_devices(self, now: datetime | None = None) -> None:
                     "WANIPConn1",
                     "ForceTermination",
                 )
+                return
+
+            if service_call.service == SERVICE_CLEANUP and mode == "stale":
+                device_hosts_list: list = await self.hass.async_add_executor_job(
+                    self.fritz_hosts.get_hosts_info
+                )
+                ha_hosts_list: ValuesView[
+                    FritzDevice
+                ] = await self.hass.async_add_executor_job(
+                    lambda: self.devices.values()  # pylint: disable=unnecessary-lambda
+                )
+
         except (FritzServiceError, FritzActionError) as ex:
             raise HomeAssistantError("Service or parameter unknown") from ex
         except FritzConnectionException as ex:
             raise HomeAssistantError("Service not supported") from ex
 
+        if mode == "stale":
+            dev_found: bool = False
+            dev_obj = None
+            for ha_host in ha_hosts_list:
+                for device in device_hosts_list:
+                    if (
+                        ha_host.hostname == device["name"]
+                        and ha_host.mac_address == device["mac"]
+                    ):
+                        dev_found = True
+                        dev_obj = device
+                        break
+                if not dev_found and dev_obj:
+                    _LOGGER.warning(
+                        "Device %s [%s] deleted",
+                        ha_host.hostname,
+                        ha_host.mac_address,
+                    )
+                    entity_reg = (
+                        await self.hass.helpers.entity_registry.async_get_registry()
+                    )
+                    entity_reg.async_remove(dev_obj.entity_id)
+                else:
+                    _LOGGER.debug(
+                        "Device %s [%s] present on Fritz!Box device, nothing to do",
+                        ha_host.hostname,
+                        ha_host.mac_address,
+                    )
+        elif mode == "disabled":
+            if not config_entry.pref_disable_new_entities:
+                _LOGGER.info(
+                    "System option 'Enable newly added entities' is on, nothing to do"
+                )
+
+            entity_reg = await self.hass.helpers.entity_registry.async_get_registry()
+            device_list: list[
+                DeviceEntry
+            ] = self.hass.helpers.entity_registry.async_entries_for_config_entry(
+                entity_reg, config_entry.entry_id
+            )
+
+            for device in device_list:
+                if (device.domain == DEVICE_TRACKER_DOMAIN and device.disabled) or (
+                    device.domain == DEVICE_SWITCH_DOMAIN
+                    and device.entity_id.endswith("internet")
+                ):
+                    _LOGGER.warning("Removing entity: %s", device.entity_id)
+                    entity_reg.async_remove(device.entity_id)
+
```

Comment:
```suggestion
    async def service_fritzbox(
        self, service_call: ServiceCall, config_entry: ConfigEntry
    ) -> None:
        """Define FRITZ!Box services."""
        _LOGGER.debug("FRITZ!Box router: %s", service_call.service)

        if not self.connection:
            raise HomeAssistantError("Unable to establish a connection")

        mode = service_call.data["mode"]
        try:
            if service_call.service == SERVICE_REBOOT:
                await self.hass.async_add_executor_job(
                    self.connection.call_action, "DeviceConfig1", "Reboot"
                )
                return

            if service_call.service == SERVICE_RECONNECT:
                await self.hass.async_add_executor_job(
                    self.connection.call_action,
                    "WANIPConn1",
                    "ForceTermination",
                )
                return

        except (FritzServiceError, FritzActionError) as ex:
            raise HomeAssistantError("Service or parameter unknown") from ex
        except FritzConnectionException as ex:
            raise HomeAssistantError("Service not supported") from ex

        if service_call.service == SERVICE_CLEANUP:
            entity_reg = await self.hass.helpers.entity_registry.async_get_registry()
            device_list: list[
                DeviceEntry
            ] = self.hass.helpers.entity_registry.async_entries_for_config_entry(
                entity_reg, config_entry.entry_id
            )

        if mode == "stale":
            device_hosts_list: list = await self.hass.async_add_executor_job(
                self.fritz_hosts.get_hosts_info
            )
            ha_hosts_list = {}
            for entity in device_list:
                if entity.unique_id.endswith('_tracker'):
                    ha_hosts_list[entity.entity_id] = entity.unique_id.replace('_tracker', '')
                elif entity.unique_id.endswith('_internet'):
                    ha_hosts_list[entity.entity_id] = entity.unique_id.replace('_internet', '')

            for ha_host_id, ha_host_mac in ha_hosts_list.items():
                dev_found: bool = False
                for device in device_hosts_list:
                    if (
                        ha_host_mac == device["mac"]
                    ):
                        dev_found = True
                        break
                if not dev_found:
                    _LOGGER.warning(
                        "Device %s deleted",
                        ha_host_id,
                    )
                    entity_reg.async_remove(ha_host_id)
                else:
                    _LOGGER.debug(
                        "Device %s present on Fritz!Box device, nothing to do",
                        ha_host_id,
                    )
        elif mode == "disabled":
            if not config_entry.pref_disable_new_entities:
                _LOGGER.info(
                    "System option 'Enable newly added entities' is on, nothing to do"
                )

            for device in device_list:
                if (device.domain == DEVICE_TRACKER_DOMAIN and device.disabled) or (
                    device.domain == DEVICE_SWITCH_DOMAIN
                    and "internet_access" in device.entity_id
                ):
                    _LOGGER.warning("Removing entity: %s", device.entity_id)
                    entity_reg.async_remove(device.entity_id)
```

Reply:
Completely reworked the code, please recheck new implementation